src/stable_hasher.rs

```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
 /// hashing and the architecture dependent `isize` and `usize` types are
 /// extended to 64 bits if needed.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+/// use std::hash::Hasher;
+///
+/// struct Hash128((u64, u64));
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+///
+/// let mut hasher = StableHasher::new();
+/// hasher.write_usize(0xFA);
+///
+/// let hash: Hash128 = hasher.finish();
+/// ```
+#[must_use]
 pub struct StableHasher {
     state: SipHasher128,
 }
 
-impl fmt::Debug for StableHasher {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:?}", self.state)
-    }
+/// Trait for retrieving the result of the stable hashing operation.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+///
+/// struct Hash128((u64, u64));
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+/// ```
+pub trait StableHasherResult: Sized {
+    /// Retriving the [`finalize`]d state of the [`StableHasher`] and construct
+    /// an [`Self`] containing the hash.
+    ///
+    /// [`finalize`]: StableHasher::finalize
+    fn finish(hasher: StableHasher) -> Self;
 }
 
 impl StableHasher {
+    /// Creates a new [`StableHasher`].
+    ///
+    /// To be used with the [`Hasher`] implementation and [`StableHasher::finish`].
     #[inline]
+    #[must_use]
     pub fn new() -> Self {
         StableHasher {
             state: SipHasher128::new_with_keys(0, 0),
         }
     }
 
+    /// Returns the typed-hash value for the values written.
+    ///
+    /// The resulting typed-hash value is constructed from an
+    /// [`StableHasherResult`] implemenation.
+    ///
+    /// To be used in-place of [`Hasher::finish`].
     #[inline]
+    #[must_use]
     pub fn finish<W: StableHasherResult>(self) -> W {
         W::finish(self)
     }
 
+    /// Returns the hash value for the values written so far.
+    ///
+    /// This method returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// See [`StableHasher::finish`] for an idomatic use of this function.
     #[inline]
+    #[must_use]
     pub fn finalize(self) -> (u64, u64) {
         self.state.finish128()
     }
 }
 
+impl fmt::Debug for StableHasher {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.state)
+    }
+}
+
 impl Hasher for StableHasher {
+    /// <div class="warning">
+    ///
+    /// Do not use this function, it will unconditionnaly panic.
+    ///
+    /// Use instead [`StableHasher::finish`], or [`StableHasher::finalize`]
+    /// which returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// </div>
     fn finish(&self) -> u64 {
```

Comment:
Does `deprecated` work on implementations? If so, you could add it here.

Reply:
Unfortunately no,
```
error: this `#[deprecated]` annotation has no effect
   --> src/stable_hasher.rs:118:5
    |
118 |     #[deprecated = "use StableHasher::finalize instead"]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove the unnecessary deprecation attribute
    |
    = note: `#[deny(useless_deprecated)]` on by default

error: could not compile `rustc-stable-hash` (lib) due to 1 previous error
```

Reply:
Afaik, the panic is from a time when there was no `StableHashResult` yet. It should be fine to just return one half of the hash or do something like [Fingerprint::combine](https://github.com/rust-lang/rust/blob/1aaab8b9f8dc488cadc4f083b3a11fb11b45cb77/compiler/rustc_data_structures/src/fingerprint.rs#L44-L51).